### DISCUSSION TITLE
Get URL on server components?

### DISCUSSION THREAD

**Discussion by onursagir** [vercel/next.js#52496] [App Router]
I wan't to create a breadcrumbs component in the a `layout.tsx` some of my pages are dynamic while others are static for example

```
| - layout.tsx
| - example
|   - page.tsx
| - [slug]
|   - page.tsx
```

Sadly the layout component will not contain the slug param unless placed into the `[slug]` folder, wanting to prevent duplicate code I think it would be ideal to access the url server-side allowing me to split it and construct a breadcrumbs component accordingly. Client component are sadly not a real option either as I need to fetch stuff from the database (not using fetch) depending on the url.

**Comment by SamKomesarook** [comment]
One solution could be to add a [middleware](https://nextjs.org/docs/app/building-your-application/routing/middleware) file with the following that grabs the request URL:
```typescript
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'
 
export function middleware(request: NextRequest) {
  // Setting cookies on the response using the `ResponseCookies` API
  const response = NextResponse.next()
  response.cookies.set('url', request.url)
  response.cookies.set({
    name: 'vercel',
    value: 'fast',
    path: '/',
  })
  return response
}
```

then access it in the layout file with [`const cookie = cookies().get("url")?.value`](https://nextjs.org/docs/app/api-reference/functions/cookies#cookiesgetname).

**Reply by ciruz** [reply]
This will disable SSG:

> Using [cookies()](https://nextjs.org/docs/app/api-reference/functions/cookies) or [headers()](https://nextjs.org/docs/app/api-reference/functions/headers) in a Server Component will opt the whole route into dynamic rendering at request time.

Related: https://nextjs.org/docs/app/building-your-application/rendering/static-and-dynamic-rendering#dynamic-functions

**Reply by onursagir** [reply]
Out of curiosity Does anyone know the specific reason why the request object hasn't been made available in server components?

**Reply by ciruz** [reply]
@onursagir yes, if you are interested: https://github.com/vercel/next.js/issues/43704
